packages/components/src/inline-edit/hook.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * External dependencies
+ */
+import { isUndefined, negate, noop, flow } from 'lodash';
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useState, useRef } from '@wordpress/element';
+
+/**
+ * @param {import("react").ChangeEvent<HTMLInputElement>} event
+ */
+const cancelEvent = ( event ) => (
+	event.preventDefault(), event.stopPropagation(), event
+);
+
+/**
+ * @param {import("react").ChangeEvent<HTMLInputElement>} event
+ */
+const getEventValue = ( { target: { value } } ) => value;
+
+/**
+ * @param {Function[]} handlers
+ */
+const mergeEvent = ( ...handlers ) =>
+	/**
+	 * @param {import("react").ChangeEvent<HTMLInputElement>} event
+	 */
+	( event ) => handlers.forEach( ( handler = noop ) => handler( event ) );
+
+/**
+ * @typedef Props
+ * @property {(value: string) => boolean} validate predicate
+ * @property {Function} onWrongInput called when a validate predicate fails
+ * @property {Function} onCommit called on enter/blur
+ * @property {string} value input value
+ */
+
+/**
+ * @param {Props} props
+ */
+export default function useInlineEdit( {
+	validate = negate( isUndefined ),
+	onWrongInput = noop,
+	onCommit = noop,
+	value: propValue,
+} ) {
+	const [ isInEditMode, setIsInEditMode ] = useState( false );
+	const [ editingValue, setEditingValue ] = useState( propValue );
+	/** @type {import('react').MutableRefObject<HTMLInputElement | undefined>} */
+	const inputRef = useRef();
+	/** @type {import('react').MutableRefObject<HTMLInputElement | undefined>} */
+	const toggleRef = useRef();
```

Comment:
Something I just learned today, it's better to use `null` as the default instead of `undefined` as `RefObject` already allows `null` under the hood. We can also just use `RefObject` instead of `MutableRefObject` as we're not mutating the ref directly, we're passing it as props to a component: https://github.com/Automattic/wp-calypso/pull/51059#issue-592875975

```suggestion
	/** @type {import('react').RefObject<HTMLInputElement>} */
	const inputRef = useRef( null );
	/** @type {import('react').RefObject<HTMLInputElement>} */
	const toggleRef = useRef( null );
```